src/Analyzers/VisualBasic/CodeFixes/RemoveUnnecessaryByVal/VisualBasicRemoveUnnecessaryByValCodeFixProvider.vb

```diff
@@ -0,0 +1,55 @@
+﻿' Licensed to the .NET Foundation under one or more agreements.
+' The .NET Foundation licenses this file to you under the MIT license.
+' See the LICENSE file in the project root for more information.
+
+Imports System.Collections.Immutable
+Imports System.Composition
+Imports System.Diagnostics.CodeAnalysis
+Imports System.Threading
+Imports Microsoft.CodeAnalysis.CodeActions
+Imports Microsoft.CodeAnalysis.CodeFixes
+Imports Microsoft.CodeAnalysis.Diagnostics
+Imports Microsoft.CodeAnalysis.Editing
+Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
+
+Namespace Microsoft.CodeAnalysis.VisualBasic.RemoveUnnecessaryByVal
+    <ExportCodeFixProvider(LanguageNames.VisualBasic), [Shared]>
+    Friend Class VisualBasicRemoveUnnecessaryByValCodeFixProvider
+        Inherits SyntaxEditorBasedCodeFixProvider
+
+        <ImportingConstructor>
+        <SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification:="Used in test code: https://github.com/dotnet/roslyn/issues/42814")>
+        Public Sub New()
+        End Sub
+
+        Public Overrides ReadOnly Property FixableDiagnosticIds As ImmutableArray(Of String) =
+            ImmutableArray.Create(IDEDiagnosticIds.RemoveUnnecessaryByValDiagnosticId)
+
+        Friend Overrides ReadOnly Property CodeFixCategory As CodeFixCategory = CodeFixCategory.CodeStyle
+
+        Public Overrides Function RegisterCodeFixesAsync(context As CodeFixContext) As Task
+            For Each diagnostic In context.Diagnostics
+                context.RegisterCodeFix(New MyCodeAction(
+                    VisualBasicAnalyzersResources.Remove_ByVal,
+                    Function(ct) FixAsync(context.Document, diagnostic, ct)),
+                    context.Diagnostics)
+            Next
+            Return Task.CompletedTask
+        End Function
+
+        Protected Overrides Async Function FixAllAsync(document As Document, diagnostics As ImmutableArray(Of Diagnostic), editor As SyntaxEditor, cancellationToken As CancellationToken) As Task
+            Dim root = Await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(False)
+            Dim node = DirectCast(root.FindNode(diagnostics(0).AdditionalLocations(0).SourceSpan), ParameterSyntax)
```

Comment:
This needs to loop through all diagnostics, not just the fix the first one.

Reply:
@mavasani, I'll try this. But I think the `diagnostics` will only contain **one** diagnostic.

I call the `FixAsync` method in the register codefix:

https://github.com/dotnet/roslyn/blob/9625a14e7405d7a3ea1dc77957d5293850b69cc8/src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/SyntaxEditorBasedCodeFixProvider.cs#L23-L27

Which in turn calls an overload of FixAllAsync with only one diagnostic:

https://github.com/dotnet/roslyn/blob/9625a14e7405d7a3ea1dc77957d5293850b69cc8/src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/SyntaxEditorBasedCodeFixProvider.cs#L29-L35

Reply:
After debugging, it seems that FixAllAsync sometimes gets called with the two diagnostics (which is surprising to me). Let me check if adding the loop fixes the problem.

Reply:
Loop in FixAllAsync is absolutely required- it does the actual fix all diagnostics in a document. Loop in RegisterCodeFixes async is a defensive approach in case you have multiple diagnostics with the same span (can sometimes happen when user has multiple versions of analyzer packages or one VSIX and one NuGet install and such. I would retain both loops.